lib/rex/proto/http/web_socket.rb

```diff
@@ -1,428 +1,445 @@
+            # This provides a hook point that is called when data is written to the WebSocket peer. Subclasses can intercept and
+            # process the data. The default functionality does nothing.
+            #
+            # @param [String] data the data that is being written
+            # @return [String, nil] if a string is returned, it's passed through the channel
+            def on_data_write(data)
+              data
+            end
+          end
 
-    #
-    # This provides a hook point that is called when data is written to the WebSocket peer. Subclasses can intercept and
-    # process the data. The default functionality does nothing.
-    #
-    # @param [String] data the data that is being written
-    # @return [String, nil] if a string is returned, it's passed through the channel
-    def on_data_write(data)
-      data
-    end
-  end
+          #
+          # Send a WebSocket::Frame to the peer.
+          #
+          # @param [WebSocket::Frame] frame the frame to send to the peer.
+          def put_wsframe(frame, opts = {})
+            put(frame.to_binary_s, opts = opts)
+          end
 
-  #
-  # Send a WebSocket::Frame to the peer.
-  #
-  # @param [WebSocket::Frame] frame the frame to send to the peer.
-  def put_wsframe(frame, opts={})
-    put(frame.to_binary_s, opts=opts)
-  end
+          #
+          # Build a WebSocket::Frame representing the binary data and send it to the peer.
+          #
+          # @param [String] value the binary value to use as the frame payload.
+          def put_wsbinary(value, opts = {})
+            put_wsframe(Frame.from_binary(value), opts = opts)
+          end
 
-  #
-  # Build a WebSocket::Frame representing the binary data and send it to the peer.
-  #
-  # @param [String] value the binary value to use as the frame payload.
-  def put_wsbinary(value, opts={})
-    put_wsframe(Frame.from_binary(value), opts=opts)
-  end
+          #
+          # Build a WebSocket::Frame representing the text data and send it to the peer.
+          #
+          # @param [String] value the binary value to use as the frame payload.
+          def put_wstext(value, opts = {})
+            put_wsframe(Frame.from_text(value), opts = opts)
+          end
 
-  #
-  # Build a WebSocket::Frame representing the text data and send it to the peer.
-  #
-  # @param [String] value the binary value to use as the frame payload.
-  def put_wstext(value, opts={})
-    put_wsframe(Frame.from_text(value), opts=opts)
-  end
+          #
+          # Read a WebSocket::Frame from the peer.
+          #
+          # @return [WebSocket::Frame] the frame that was received from the peer.
+          def get_wsframe(_opts = {})
+            frame = Frame.new
+            frame.header.read(self)
+            payload_data = ''
+            while payload_data.length < frame.payload_len
+              chunk = read(frame.payload_len - payload_data.length)
+              return nil if chunk.nil? # no partial reads!
```

Comment:
Do you know what can cause the partial reads?
Would it be worth logging out that one happened?